// agile-sdk/dist/idm/user.js
/*******************************************************************************
 *Copyright (C) 2017 FBK.
 *All rights reserved. This program and the accompanying materials
 *are made available under the terms of the Eclipse Public License v1.0
 *which accompanies this distribution, and is available at
 *http://www.eclipse.org/legal/epl-v10.html
 *
 *Contributors:
 *    FBK - initial API and implementation
 ******************************************************************************/
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _axios = require('axios');

var _axios2 = _interopRequireDefault(_axios);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var idm = function idm(base, token) {
  base = '' + base;
  var instance = _axios2.default.create({
    headers: { 'Authorization': 'bearer ' + token }
  });

  return {
    /**
    * @summary Get the user information for the user currently logged in, i.e. token provided when agileSDK was created
    * @name getCurrentUserInfo
    * @public
    * @function
    * @memberof agile.idm.user
    * @fulfil {Object} userInfo - object with user information
    * @returns {Promise}
    * @example
    * agile.idm.user.getCurrentUserInfo().then(function(info) {
    *  console.log(info);
    * });
    **/
    getCurrentUserInfo: function getCurrentUserInfo() {
      return instance.request({
        method: 'GET',
        url: base + '/oauth2/api/userinfo'
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Show information for a particular user by username and authentication type
    * @name get
    * @public
    * @function
    * @memberof agile.idm.user
    * @param {String} userName user name
    * @param {String} authType authentication type
    * @fulfil {Object} user found
    * @returns {Promise}
    * @example
    * agile.idm.user.get('alice','agile-local').then(function(user) {
    *   console.log(user);
    * });
    **/
    get: function get(userName, authType) {
      return instance.request({
        method: 'GET',
        url: base + '/api/v1/user/',
        params: { 'auth_type': authType, 'user_name': userName }
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Create user
    * @name create
    * @public
    * @function
    * @memberof agile.idm.user
    * @param {object} including userName user name
    * @param {String} authType authentication type
    * @param {Object} [options] continaing  role  of the user as 'role' and password as 'password'
    * @fulfil {Object} user created
    * @returns {Promise}
    * @example
    * agile.idm.user.create('bob','agile-local',{'role':'admin', 'password':'secret'}).then(function(user) {
    *   console.log('user created!'+user);
    * });
    **/
    create: function create(userName, authType, options) {
      var user = {
        'auth_type': authType,
        'user_name': userName
      };
      if (options && options.role) {
        user.role = options.role;
      }
      if (options && options.password) {
        user.password = options.password;
      }
      return instance.request({
        method: 'POST',
        url: base + '/api/v1/user/',
        data: user
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Delete a user
    * @name delete
    * @public
    * @function
    * @memberof agile.idm.user
    * @param {String} userName user name
    * @param {String} authType authentication type
    * @fulfil {Undefined}
    * @returns {Promise}
    * @example
    * agile.idm.user.delete('bob','agile-local').then(function() {
    *   console.log('user removed!');
    * });
    **/
    delete: function _delete(userName, authType) {
      return instance.request({
        method: 'DELETE',
        url: base + '/api/v1/user/',
        params: { 'auth_type': authType, 'user_name': userName }
      }).then(function (res) {
        return res;
      }).catch(_utils.errorHandler);
    }

  };
};

exports.default = idm;
