// agile-sdk/dist/protocol/index.js
/*******************************************************************************
 *Copyright (C) 2017 FBK.
 *All rights reserved. This program and the accompanying materials
 *are made available under the terms of the Eclipse Public License v1.0
 *which accompanies this distribution, and is available at
 *http://www.eclipse.org/legal/epl-v10.html
 *
 *Contributors:
 *    FBK - initial API and implementation
 ******************************************************************************/
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _axios = require('axios');

var _axios2 = _interopRequireDefault(_axios);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var protocol = function protocol(base) {
  base = base + '/protocol';
  return {
    /**
    * @summary Disconnect from the device
    * @name disconnect
    * @public
    * @function
    * @memberof agile.protocol
    * @param {String} protocolId - Agile Id of the protocol
    * @param {String} deviceId - Agile internal Id of the protocol
    * @fulfil {Undefined}
    * @returns {Promise}
    *
    * @example
    * agile.protocol.disconnect('Bluetooth LE', 'bleB0B448BE5084').then(function() {
    *  console.log('Disconnected!');
    * });
    **/
    disconnect: function disconnect(protocolId, deviceId) {
      return (0, _axios2.default)({
        method: 'POST',
        url: base + '/' + protocolId + '/connection/' + deviceId
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Connect to the device
    * @name connect
    * @public
    * @function
    * @memberof agile.protocol
    * @param {String} protocolId - Agile Id of the protocol
    * @param {String} deviceId - Agile internal Id of the protocol
    * @fulfil {Undefined}
    * @returns {Promise}
    *
    * @example
    * agile.protocol.connect('Bluetooth LE', 'bleB0B448BE5084').then(function() {
    *  console.log('Connected!');
    * });
    **/
    connect: function connect(protocolId, deviceId) {
      return (0, _axios2.default)({
        method: 'DELETE',
        url: base + '/' + protocolId + '/connection/' + deviceId
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Call a read via protocol
    * @name read
    * @public
    * @function
    * @memberof agile.protocol
    * @param {String} protocolId - Agile Id of the protocol
    * @param {String} deviceId - Agile internal Id of the protocol
    * @fulfil {Object}
    * @returns {Promise}
    *
    * @example
    * agile.protocol.read('Bluetooth LE', 'bleB0B448BE5084').then(function(data) {
    *  console.log(data);
    * });
    **/
    read: function read(protocolId, deviceId) {
      return (0, _axios2.default)({
        url: base + '/' + protocolId + '/' + deviceId
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Call a write via protocol
    * @name write
    * @public
    * @function
    * @memberof agile.protocol
    * @param {String} protocolId - Agile Id of the protocol
    * @param {String} deviceId - Agile internal Id of the protocol
    * @param {object} data - An object containing the content to write
    * @fulfil {Undefined}
    * @returns {Promise}
    *
    * @example
    * agile.protocol.write('Bluetooth LE', 'bleB0B448BE5084', data).then(function() {
    *  console.log('wrote data!');
    * });
    **/
    write: function write(protocolId, deviceId, data) {
      return (0, _axios2.default)({
        method: 'POST',
        url: base + '/' + protocolId + '/' + deviceId,
        data: data
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    }
  };
};

exports.default = protocol;
