// agile-sdk/dist/deviceManager/index.js
/*******************************************************************************
 *Copyright (C) 2017 FBK.
 *All rights reserved. This program and the accompanying materials
 *are made available under the terms of the Eclipse Public License v1.0
 *which accompanies this distribution, and is available at
 *http://www.eclipse.org/legal/epl-v10.html
 *
 *Contributors:
 *    FBK - initial API and implementation
 ******************************************************************************/
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _axios = require('axios');

var _axios2 = _interopRequireDefault(_axios);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var deviceManager = function deviceManager(base) {
  base = base + '/devices';
  return {
    /**
    * @summary Get all or single device definition
    * @name get
    * @public
    * @function
    * @memberof agile.deviceManager
    * @param {String} [deviceId] - Agile device Id
    * @fulfil {Array} - devices
    * @returns {Promise}
    * @example
    * agile.deviceManager.get('bleB0B448BE5084').then(function(device) {
    *   console.log(device);
    * });
    **/
    get: function get(deviceId) {
      var url = '' + base;
      if (deviceId) {
        url = base + '/' + deviceId;
      }
      return (0, _axios2.default)({
        method: 'GET',
        url: url
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Delete a device definition and unregister it
    * @name delete
    * @public
    * @function
    * @memberof agile.deviceManager
    * @param {String} deviceId - Agile device Id
    * @fulfil {undefined}
    * @returns {Promise}
    * @example
    * agile.deviceManager.delete('bleB0B448BE5084').then(function() {
    *   console.log('Device bleB0B448BD1085 deleted');
    * });
    **/
    delete: function _delete(deviceId) {
      return (0, _axios2.default)({
        method: 'DELETE',
        url: base + '/' + deviceId
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Register a new device based on information from ProtocolManager and device type
    * @name create
    * @public
    * @function
    * @memberof agile.deviceManager
    * @param {Object} deviceOverview
    * @param {string} type
    * @fulfil {Object} - device
    * @returns {Promise}
    * @example
    * const deviceOverview = {
    *   "name": "CC2650 SensorTag",
    *   "protocol": "iot.agile.protocol.BLE",
    *   "id": "B0:B4:48:BD:10:85",
    *   "status": "CONNECTED"
    * };
    * const type = "TI SensorTag";
    *
    * agile.deviceManager.create(deviceOverview, type).then(function(newDevice) {
    *  console.log(newDevice);
    * });
    **/
    create: function create(deviceOverview, type) {
      return (0, _axios2.default)({
        method: 'POST',
        url: '' + base,
        data: {
          overview: deviceOverview,
          type: type
        }
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Get matching types for a device overview
    * @name typeof
    * @public
    * @function
    * @memberof agile.deviceManager
    *
    * @param {Object} deviceOverview
    * @fulfil {Array} - deviceTypes
    * @returns {Promise}
    * @example
    * const deviceOverview = {
    *   "name": "CC2650 SensorTag",
    *   "protocol": "iot.agile.protocol.BLE",
    *   "id": "B0:B4:48:BD:10:85",
    *   "status": "CONNECTED"
    * };
    *
    * agile.deviceManager.typeof(deviceOverview).then(function(deviceTypes) {
    *  console.log(deviceTypes);
    * });
    **/
    typeof: function _typeof(deviceOverview) {
      return (0, _axios2.default)({
        method: 'POST',
        url: base + '/typeof',
        data: deviceOverview
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    }
  };
};

exports.default = deviceManager;
