// agile-sdk/dist/protocolManager/discovery.js
/*******************************************************************************
 *Copyright (C) 2017 FBK.
 *All rights reserved. This program and the accompanying materials
 *are made available under the terms of the Eclipse Public License v1.0
 *which accompanies this distribution, and is available at
 *http://www.eclipse.org/legal/epl-v10.html
 *
 *Contributors:
 *    FBK - initial API and implementation
 ******************************************************************************/
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _axios = require('axios');

var _axios2 = _interopRequireDefault(_axios);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var discovery = function discovery(base) {
  var protocolBase = base.slice(0, -1);
  base = base + '/discovery';
  return {
    /**
    * @summary Start device discovery on all or single protocol
    * @name start
    * @public
    * @function
    * @memberof agile.protocolManager.discovery
    * @param [protocolId] - Agile protocol Id
    * @fulfil {null}
    * @returns {Promise}
    *
    * @example
    * agile.protocolManager.discovery.start().then(function() {
    *   console.log('All protocols discovery is on');
    * });
    * agile.protocolManager.discovery.start('Bluetooth LE').then(function() {
    *   console.log('Bluetooth LE protocols discovery is on');
    * });
    **/
    start: function start(protocolId) {
      var url = base;
      if (protocolId) {
        url = protocolBase + '/' + protocolId + '/discovery';
      }
      return (0, _axios2.default)({
        method: 'POST',
        url: url
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Stop device discovery on all or single protocol
    * @name stop
    * @public
    * @function
    * @memberof agile.protocolManager.discovery
    *
    * @param [protocolId] - Agile protocol Id
    * @fulfil {null}
    * @returns {Promise}
    *
    * @example
    * agile.protocolManager.discovery.stop().then(function() {
    *   console.log('All protocols discovery is off');
    * });
    * agile.protocolManager.discovery.stop('Bluetooth LE').then(function() {
    *   console.log('Bluetooth LE discovery is off');
    * });
    **/
    stop: function stop(protocolId) {
      var url = base;
      if (protocolId) {
        url = protocolBase + '/' + protocolId + '/discovery';
      }
      return (0, _axios2.default)({
        method: 'DELETE',
        url: url
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Return the status of discovery on the all or single protocol
    * @name status
    * @public
    * @function
    * @memberof agile.protocolManager.discovery
    *
    * @param [protocolId] - Agile protocol Id
    * @fulfil {Object}
    * @returns {Promise}
    *
    * @example
    * agile.protocolManager.discovery.status().then(function(status) {
    *   console.log(status);
    * });
    * agile.protocolManager.discovery.status('Bluetooth LE').then(function(status) {
    *   console.log(status);
    * });
    **/
    status: function status(protocolId) {
      var url = base;
      if (protocolId) {
        url = protocolBase + '/' + protocolId + '/discovery';
      }
      return (0, _axios2.default)({
        method: 'GET',
        url: url
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    }
  };
};

exports.default = discovery;
