// agile-sdk/dist/protocolManager/index.js
/*******************************************************************************
 *Copyright (C) 2017 FBK.
 *All rights reserved. This program and the accompanying materials
 *are made available under the terms of the Eclipse Public License v1.0
 *which accompanies this distribution, and is available at
 *http://www.eclipse.org/legal/epl-v10.html
 *
 *Contributors:
 *    FBK - initial API and implementation
 ******************************************************************************/
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _axios = require('axios');

var _axios2 = _interopRequireDefault(_axios);

var _discovery = require('./discovery');

var _discovery2 = _interopRequireDefault(_discovery);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var protocolManager = function protocolManager(base) {
  base = base + '/protocols';
  return {
    /**
    * @summary Get the list of registered protocols
    * @name get
    * @public
    * @function
    * @memberof agile.protocolManager
    * @fulfil {Array} - protocols
    * @returns {Promise}
    * @example
    * agile.protocolManager.protocols.get().then(function(protocols) {
    *   console.log(protocols);
    * });
    **/
    get: function get() {
      return (0, _axios2.default)({
        method: 'GET',
        url: '' + base
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Unregister a Dbus Protocol object reference
    * @name delete
    * @public
    * @function
    * @memberof agile.protocolManager
    *
    * @param {String} protocolId - Agile protocol Id
    * @fulfil {null}
    * @returns {Promise}
    *
    * @example
    * agile.protocolManager.protocols.delete(protocolId).then(function() {
    *   console.log('protocol has been unregistered');
    * });
    **/
    delete: function _delete() {
      return (0, _axios2.default)({
        method: 'DELETE',
        url: '' + base
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary Register a new Dbus object implementing the protocol API
    * @name create
    * @public
    * @function
    * @memberof agile.protocolManager
    *
    * @param {String} protocolId - Agile protocol Id
    * @fulfil {null}
    * @returns {Promise}
    *
    * @example
    * agile.protocolManager.protocols.create(protocolId).then(function() {
    *   console.log('protocol has been registered');
    * });
    **/
    status: function status() {
      return (0, _axios2.default)({
        method: 'POST',
        url: '' + base
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @summary List all discovered devices on all available protocols
    * @name devices
    * @public
    * @function
    * @memberof agile.protocolManager
    *
    * @fulfil {Array} - devices
    * @returns {Promise}
    *
    * @example
    * agile.protocolManager.devices().then(function(devices) {
    *  console.log(devices);
    * });
    **/
    devices: function devices() {
      return (0, _axios2.default)({
        url: base + '/devices'
      }).then(function (res) {
        return res.data;
      }).catch(_utils.errorHandler);
    },
    /**
    * @namespace discovery
    * @memberof agile.protocolManager
    **/
    discovery: (0, _discovery2.default)(base)
  };
};

exports.default = protocolManager;
